README.md
# Clojure-Schemas

hospital-sales/CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file. This change log follows the conventions of [keepachangelog.com](http://keepachangelog.com/).

## [Unreleased]
### Changed
- Add a new arity to `make-widget-async` to provide a different widget shape.

## [0.1.1] - 2022-09-26
### Changed
- Documentation on how to make the widgets.

### Removed
- `make-widget-sync` - we're all async, all the time.

### Fixed
- Fixed widget maker to keep working when daylight savings switches over.

## 0.1.0 - 2022-09-26
### Added
- Files from the new template.
- Widget maker public API - `make-widget-sync`.

[Unreleased]: https://github.com/your-name/hospital-sales/compare/0.1.1...HEAD
[0.1.1]: https://github.com/your-name/hospital-sales/compare/0.1.0...0.1.1


hospital-sales/test/hospital_sales/core_test.clj
(ns hospital-sales.core-test
  (:require [clojure.test :refer :all]
            [hospital-sales.core :refer :all]))

(deftest a-test
  (testing "FIXME, I fail."
    (is (= 0 1))))


hospital-sales/src/hospital_sales/aula2.clj
(ns hospital_sales.aula2
  (:use clojure.pprint)
  (:require [schema.core :as s]))

(s/set-fn-validation! true)

;(s/defrecord Paciente
;[id :- Long, nome :- s/Str])

;(pprint (Paciente. 26 "Jean"))
; O defrecord é expansivo, aceita mais valores do que esperado
;(pprint (map->Paciente {26 "Jean"}))
;(pprint (map->Paciente {26 "Jean"}))

(def Paciente
  "Schema de um paciente"
  {:id s/Num, :nome s/Str})
; Explica o Schema
(pprint (s/explain Paciente))

(pprint (s/validate Paciente {:id 26, :nome "jean"}))
(pprint (s/validate Paciente {:id 02, :nome "marina"}))
(pprint (s/validate Paciente {:id 19, :nome "renata"}))

; Typo é pego pelo schema, esse modelo de erro seria pego em testes automatizados.
;(pprint (s/validate Paciente {:id 19, :name "renata"})) ;name não é uma chave.
; A questão é se você deseja que seja forward compatible ou não. Entender o custo/beneficio
;Forward - Compativel para novas informações

;Reclama um erro por naão estar com as chaves necessaárias.
;(pprint (s/validate Paciente {:id 19}))

(s/defn novo-paciente :- Paciente
  [id :- s/Num, nome :- s/Str]
  {:id id, :nome nome})

(pprint (novo-paciente 19 "Minerva"))

(defn estritamente-positivo? [x]
  (> x 0))

;(def EstritamentePositivo (s/pred estritamente-positivo?))
(def EstritamentePositivo (s/pred estritamente-positivo? 'estritamente-positivo'))

(pprint (s/validate EstritamentePositivo 15))

(def Paciente
  "Schema de um paciente"
  {:id (s/constrained s/Int pos?), :nome s/Str})

;(def Paciente
;  "Schema de um paciente"
;  {:id (s/constrained s/Int estritamente-positivo?), :nome s/Str})

(def Paciente
  "Schema de um paciente"
  {:id (s/constrained s/Int #(> % 0) 'inteiro-estritamente-positivo), :nome s/Str})

(pprint (s/validate Paciente {:id -02, :nome "marina"}))

;Podemos usar predicates com Schemas para Definir regras extras que limitam os valores de nossos dados. Cada predicado limita os valores de um dado.

hospital-sales/src/hospital_sales/aula3.clj
(ns hospital_sales.aula3
  (:use clojure.pprint)
  (:require [schema.core :as s]))

(s/set-fn-validation! true)

 (def PosInt (s/pred pos-int? 'inteiro-positivo))
(def Paciente
  {:id PosInt, :nome s/Str})

(s/defn novo-paciente :- Paciente
  [id :- PosInt,
   nome :- s/Str]
  {:id id, :nome nome})

(pprint (novo-paciente 02, "Marina"))
;(pprint (novo-paciente -3, "Marina"))

(defn Maior-ou-igual-a-zero? [x] (>= x 0))
(def ValorFinanceiro (s/constrained s/Num Maior-ou-igual-a-zero?))

(def Pedido
  {:paciente Paciente
   :valor ValorFinanceiro
   :procedimento s/Keyword})

(s/defn novo-pedido :- Pedido
  [paciente :- Paciente,
   valor :- ValorFinanceiro,
   procedimento :- s/Keyword]
  {
      :paciente     paciente,
      :valor        valor,
      :procedimento procedimento
   })
(pprint (novo-pedido (novo-paciente 02, "marina"), 300, :raio-x))
;(pprint (novo-pedido (novo-paciente 02, "marina"), -300, :raio-x))

(def Numeros [s/Num])

(pprint (s/validate Numeros [1]))
(pprint (s/validate Numeros [1, 2, 3, 4, 5 ]))
(pprint (s/validate Numeros [1, 1.5, 2, 2.5]))
(pprint (s/validate Numeros []))
(pprint (s/validate Numeros nil))e

(def Plano [s/Keyword])

(def Paciente
{:id PosInt, :nome s/Str, :plano Plano})

(pprint (s/validate Paciente {:id 19, :nome "Renata", :plano [:raio-x, :ultrassom]}))
(pprint (s/validate Paciente {:id 19, :nome "Renata", :plano [:raio-x]}))
(pprint (s/validate Paciente {:id 19, :nome "Renata", :plano []}))
(pprint (s/validate Paciente {:id 19, :nome "Renata", :plano nil}))


;Vetores com 0 ou mais elementos de um único tipo são tranquilos de modelar e lidar no dia a dia.
;Vetores que misturam tipos de acordo com posições são complicados por se assemelham a objetos ou mapas, mas mais restritos.
;Vetores por exemplo não permitem a inexistência de valores como um mapa permitiria ao representar o mesmo tipo de dado.

hospital-sales/src/hospital_sales/aula4.clj
(ns hospital_sales.aula4
  (:use clojure.pprint)
  (:require [schema.core :as s]))

(s/set-fn-validation! true)

(def PosInt (s/pred pos-int? 'inteiro-positivo))

(def Plano [s/Keyword])

(def Paciente
  {:id PosInt,
   :nome s/Str,
   :plano Plano
   ;Definir chave Opcional
  (s/optional-key :nascimnto) s/Str })


(pprint (s/validate Paciente {:id 19, :nome "Renata", :plano [:raio-x, :ultrassom]}))
(pprint (s/validate Paciente {:id 19, :nome "Renata", :plano [:raio-x]}))
(pprint (s/validate Paciente {:id 19, :nome "Renata", :plano []}))
(pprint (s/validate Paciente {:id 19, :nome "Renata", :plano nil}))

(def Pacientes
  {PosInt Paciente})

(let [renata {:id 19, :nome "Renata", :plano [:raio-x, :ultrassom]}]
  (pprint (s/validate Pacientes {19 renata} )))

;Mapas podem representar entidades, a cada novo mapa que representa um Usuario as mesmas chaves são usadas.
;Ou podem representar coleções indexadas por um tipo de chave, de forma dinâmica. A chave pode ser, por exemplo, ids.

hospital-sales/src/hospital_sales/aula5.clj
(ns hospital_sales.aula5
  (:use clojure.pprint)
  (:require [schema.core :as s]))

(s/set-fn-validation! true)

(def PosInt (s/pred pos-int? 'inteiro-positivo))

(def Plano [s/Keyword])

(def Paciente
  {:id PosInt,
   :nome s/Str,
   :plano Plano
   ;Definir chave Opcional
   (s/optional-key :nascimnto) s/Str })

(def Pacientes
  {PosInt Paciente})

(def Visitas
  {PosInt [s/Str]})

;Foi removido o  IF e o THROW, pois a validação foi ativado dentro do def Paciente
;Nil não é um Pós Int
(s/defn adiciona-paciente :- Pacientes
  [pacientes :- Pacientes, paciente :- Paciente]
  (let [id (:id paciente)]
    (assoc pacientes id paciente)
  ))

(s/defn adiciona-visita :- Visitas
  [visitas :- Visitas, paciente :- PosInt, novas-visitas :- [s/Str] ]
  (if (contains? visitas paciente)
    (update visitas paciente concat novas-visitas)
    (assoc visitas paciente novas-visitas)))

(s/defn imprime-relatorio-de-paciente
  [visitas :- Visitas, paciente :- PosInt]
  (println "Visitas do paciente" paciente "são" (get visitas paciente))
  )

(defn testa-uso-de-pacientes []
  (let [jean {:id 26, :nome "Jean", :plano []}
        marina {:id 02, :nome "Marina", :plano [:Pneumologista, :Dermatologista]}
        renata {:id 19, :nome "Renata", :plano [:Dermatologista]}
        pacientes (reduce adiciona-paciente {} [jean, marina, renata])
        visitas {}
        visitas (adiciona-visita visitas 26 ["01/01/2022"])
        visitas (adiciona-visita visitas 02 ["01/02/2022","07/02/2022"])
        visitas (adiciona-visita visitas 19 ["01/03/2022"])]
    (pprint pacientes)
    (pprint visitas)
    (imprime-relatorio-de-paciente visitas 26)))

(testa-uso-de-pacientes)


hospital-sales/src/hospital_sales/aula1.clj
(ns hospital_sales.aula1
  (:use clojure.pprint)
  (:require [schema.core :as s]))

(defn adiciona-paciente
  [pacientes, paciente]
  (if-let [id (:id paciente)]
    (assoc pacientes id paciente)
    (throw (ex-info "Paciente não possui id" {:paciente paciente}))))

(defn adiciona-visita
  [visitas, paciente, novas-visitas]
  (if (contains? visitas paciente)
    (update visitas paciente concat novas-visitas)
    (assoc visitas paciente novas-visitas)))

(defn imprime-relatorio-de-paciente
      [visitas, paciente]
  (println "Visitas do paciente" paciente "são" (get visitas paciente))
  )

; Retorna um vetor com 3 pacientes
(defn testa-uso-de-pacientes []
  (let [jean {:id 26, :nome "Jean"}
        marina {:id 02, :nome "Marina"}
        renata {:id 19, :nome "Renata"}]
    (pprint [jean marina renata])))

; Retorna um mapa com os pacientes
(defn testa-uso-de-pacientes []
  (let [jean {:id 26, :nome "Jean"}
        marina {:id 02, :nome "Marina"}
        renata {:id 19, :nome "Renata"}
        ;Adiciona em pacientes o retorno da função "adiciona-paciente"
        ;Toda vez que adiciona o paciente o reduce cria uma lista para que seja adicionado o outro
        pacientes (reduce adiciona-paciente {} [jean, marina, renata])]
    (pprint pacientes)))

(defn testa-uso-de-pacientes []
  (let [jean {:id 26, :nome "Jean"}
        marina {:id 02, :nome "Marina"}
        renata {:id 19, :nome "Renata"}
        pacientes (reduce adiciona-paciente {} [jean, marina, renata])
        visitas {}
        visitas (adiciona-visita visitas 26 ["01/01/2022"])
        visitas (adiciona-visita visitas 02 ["01/02/2022","07/02/2022"])
        visitas (adiciona-visita visitas 19 ["01/03/2022"])]
    (pprint pacientes)
    (pprint visitas)
    (imprime-relatorio-de-paciente visitas 26)))

(testa-uso-de-pacientes)

;SCHEMA serve como validador.
; S/valide = S de schema
(pprint (s/validate Long 15))

; Define que toda s/fn vai ser validado
(s/set-fn-validation! true)

(s/defn teste-simples [x]
  (pprint x))
(teste-simples "Jean")

(s/defn teste-validacao-simples [x :- Long]
  (pprint x))

(teste-validacao-simples 02)

(s/defn imprime-relatorio-de-paciente
  [visitas, paciente :- Long]
  (println "Visitas do paciente" paciente "são" (get visitas paciente))
  )
(testa-uso-de-pacientes)

(s/defn novo-paciente
  [id :- Long, nome :- s/Str]
  {:id id, :nome nome})

(pprint (novo-paciente 19 "Renata"))

;Validaçaão de schemas é principalmente usando quando:
; 1. Estão sendo realizados testes automatizados
; 2. Na borda do sistema, quando recebe info de terceiros.

;Schemas podem nos ajudar em diversas situações. Qual cenário faz sentido para mantermos nossas validações de schemas?
;Ativo em desenvolvimento e testes com boa qualidade, desativado em produção exceto nas camadas de entrada de dados externos.
;Essa abordagem faz sentido pois temos as garantias dos schemas nas bordas dos sistemas em produção e dentro do sistema em testes com qualidade.